### PR TITLE
MagicGLP (ready to merge)

### PR DESCRIPTION
**Pull request by 0xCalibur**

### CODE REVIEW

File: src/periphery/GlpVaultHarvestor.sol [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    event OutputTokenChanged(IERC20 indexed, IERC20 indexed);
+    event RewardRouterV2Changed(IGmxRewardRouterV2 indexed, IGmxRewardRouterV2 indexed);
+    error NotAllowedOperator();
+    error ReturnRewardBalance(uint256 balance);
+
+    IGmxGlpRewardHandler public immutable wrapper;
+
+    IERC20 public rewardToken;
+    IERC20 public outputToken;
+    IGmxRewardRouterV2 public rewardRouterV2;
```
Comment (0xm3rlin) [inline src/periphery/GlpVaultHarvestor.sol]: there is the need to track both old and new rewardRouter,
consider calling one mintRewardRouter and claimRewardRouter, the former is the new one the latter the old one

File: src/periphery/GlpVaultHarvestor.sol [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            rewardToken.balanceOf(address(wrapper)) +
+            IGmxRewardTracker(rewardRouterV2.feeGmxTracker()).claimable(address(wrapper)) +
+            IGmxRewardTracker(rewardRouterV2.feeGlpTracker()).claimable(address(wrapper));
+    }
+
+    function run(uint256 amountOutMin, bytes calldata data) external onlyOperators {
+        wrapper.harvest();
+        wrapper.swapRewards(amountOutMin, rewardToken, outputToken, address(this), data);
+
+        // TODO, mint GLP and transfer to GlpVault token
```
Comment (0xm3rlin) [inline src/periphery/GlpVaultHarvestor.sol]: GLP mint has to use the new GLP router

File: src/periphery/GlpVaultHarvestor.sol [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    function claimable() external view returns (uint256) {
+        return
+            IGmxRewardTracker(rewardRouterV2.feeGmxTracker()).claimable(address(wrapper)) +
+            IGmxRewardTracker(rewardRouterV2.feeGlpTracker()).claimable(address(wrapper));
+    }
+
+    function totalRewardsBalanceAfterClaiming() external view returns (uint256) {
+        return
+            rewardToken.balanceOf(address(wrapper)) +
+            IGmxRewardTracker(rewardRouterV2.feeGmxTracker()).claimable(address(wrapper)) +
```
Comment (0xm3rlin) [inline src/periphery/GlpVaultHarvestor.sol]: claimable should also be queried from the old claimRewardRouter

File: src/tokens/GmxGlpVault.sol [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+import "BoringSolidity/libraries/BoringERC20.sol";
+import "BoringSolidity/ERC20.sol";
+import "BoringSolidity/BoringOwnable.sol";
+import "interfaces/IERC20Vault.sol";
+import {GmxGlpWrapperData} from "tokens/GmxGlpWrapper.sol";
+
+/// @dev Vault version of the GLP Wrapper, auto compound yield increase shares value similar to sSPELL
+/// storage is the same as the wrapper and allow to use the same GmxGlpRewardHandler implementation.
+contract GmxGlpVault is GmxGlpWrapperData, IERC20Vault {
```
Comment (0xm3rlin) [inline src/tokens/GmxGlpVault.sol]: Do we want to be ERC4626 compatible here?

Comment (0xm3rlin) [reply src/tokens/GmxGlpVault.sol]: we could also add that functionality through upgrading the rewardRouter I would assume

File: script/GlpCauldronComp.s.sol [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            address degenBox = constants.getAddress("arbitrum.degenBox");
+            address masterContract = constants.getAddress("arbitrum.cauldronV4");
+            address weth = constants.getAddress("arbitrum.weth");
+            address glpManager = constants.getAddress("arbitrum.gmx.glpManager");
+            address rewardRouterV2 = constants.getAddress("arbitrum.gmx.rewardRouterV2");
+            address glpRewardRouter = constants.getAddress("arbitrum.gmx.glpRewardRouter");
+
+            startBroadcast();
+
+            vault = new GmxGlpVault(ERC20(sGlp), "AbracadabraStakedGlpVault", "abra-GlpVault");
```
Comment (0xm3rlin) [inline script/GlpCauldronComp.s.sol]: the new name is going to be magicGLP or mGLP
```suggestion
            vault = new GmxGlpVault(ERC20(sGlp), "magicGLP", "mGLP");
```

File: src/periphery/GlpVaultHarvestor.sol [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    function run(uint256 minGlp) external onlyOperators {
+        vault.harvest();
+        weth.safeTransferFrom(address(vault), address(this), weth.balanceOf(address(vault)));
+        weth.withdraw(weth.balanceOf(address(this)));
+
+        uint256 glpAmount = glpRewardRouter.mintAndStakeGlpETH{value: address(this).balance}(0, minGlp);
+        IERC20 asset = IERC4626(address(vault)).asset();
+        asset.safeTransfer(address(vault), glpAmount);
```
Comment (0xm3rlin) [inline src/periphery/GlpVaultHarvestor.sol]: great to make sure same asset sGLP contract is used throughout

**Comment by 0xm3rlin** [comment]
looks good 👍

**Comment by 0xm3rlin** [comment]
We still need to update the fee to 1% for mGLP?

**Comment by 0xCalibur** [comment]
Merged in main

**Pull request merged**